coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+	go func() {
+		defer close(closed)
+		defer ticker.Stop()
+		// Force an initial tick.
+		doTick(dbtime.Time(clk.Now()).UTC())
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case tick := <-ticker.C:
+				ticker.Stop()
+				doTick(dbtime.Time(tick).UTC())
+			}
+		}
+	}()
+	return &reportGenerator{
+		cancel: cancelFunc,
+		closed: closed,
+	}
+}
+
+type reportGenerator struct {
+	cancel context.CancelFunc
+	closed chan struct{}
+}
+
+func (i *reportGenerator) Close() error {
+	i.cancel()
+	<-i.closed
+	return nil
+}
+
+const failedWorkspaceBuildsReportFrequencyDays = 7
+
+func reportFailedWorkspaceBuilds(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) error {
+	now := clk.Now()
+	since := now.Add(-failedWorkspaceBuildsReportFrequencyDays * 24 * time.Hour)
+
+	statsRows, err := db.GetWorkspaceBuildStatsByTemplates(ctx, dbtime.Time(since).UTC())
+	if err != nil {
+		return xerrors.Errorf("unable to fetch failed workspace builds: %w", err)
+	}
+	sort.Slice(statsRows, func(i, j int) bool {
+		return statsRows[i].TemplateName < statsRows[j].TemplateName
+	})
+
+	reportGeneratedNow := map[uuid.UUID]bool{}
+	for _, stats := range statsRows {
+		var failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow
+		reportData := map[string]any{}
+
+		if stats.FailedBuilds > 0 {
+			failedBuilds, err = db.GetFailedWorkspaceBuildsByTemplateID(ctx, database.GetFailedWorkspaceBuildsByTemplateIDParams{
+				TemplateID: stats.TemplateID,
+				Since:      dbtime.Time(since).UTC(),
+			})
+			if err != nil {
+				logger.Error(ctx, "unable to fetch failed workspace builds", slog.F("template_id", stats.TemplateID), slog.Error(err))
+				continue
+			}
+
+			// There are some failed builds, so we have to prepare input data for the report.
+			reportData = buildDataForReportFailedWorkspaceBuilds(failedWorkspaceBuildsReportFrequencyDays, stats, failedBuilds)
+		}
+
+		templateAdmins, err := findTemplateAdmins(ctx, db, stats)
+		if err != nil {
+			logger.Error(ctx, "unable to find template admins", slog.F("template_id", stats.TemplateID), slog.Error(err))
+			continue
+		}
+
+		for _, templateAdmin := range templateAdmins {
+			reportLog, err := db.GetReportGeneratorLogByUserAndTemplate(ctx, database.GetReportGeneratorLogByUserAndTemplateParams{
+				UserID:                 templateAdmin.ID,
+				NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			})
+			if err != nil && !xerrors.Is(err, sql.ErrNoRows) { // sql.ErrNoRows: report not generated yet
+				logger.Error(ctx, "unable to get recent report generator log for user", slog.F("user_id", templateAdmin.ID), slog.Error(err))
+				continue
+			}
+
+			if !reportLog.LastGeneratedAt.IsZero() && reportLog.LastGeneratedAt.Add(failedWorkspaceBuildsReportFrequencyDays*24*time.Hour).After(now) {
+				// report generated recently, no need to send it now
+				continue
+			}
+
+			reportGeneratedNow[templateAdmin.ID] = true
```

Comment:
Should this not only be set after checking if there are failed builds?

Reply:
I think so. It indicates that some template admins will receive reports, and system should store the timestamp when it happened.

Did you have a specific use case in mind or is it just a general question?

Reply:
Well if `failedBuilds == 0` then it'lll still enter into the loop on line 181, no?

Reply:
This seems to be a naming problem, `reportGeneratedNow` indicates that the report has been generated, but in fact it means that _the job processed the template admin and the template, and should check again in a week_ (it does not matter if the report was created or not).